File: rag_new/frontend/src/utils/sourceFormatter.js
/**
 * Utility functions for formatting source displays with original path information.
 */

/**
 * Check if a file path is a temporary file path.
 * @param {string} filePath - File path to check
 * @returns {boolean} True if it's a temp file path, False otherwise
 */
export function isTempFilePath(filePath) {
    if (!filePath) return false;
    
    const tempIndicators = [
        'Temp',
        'tmp',
        '/tmp/',
        '\\temp\\',
        'tempfile',
        'temp_file'
    ];
    
    const filePathLower = filePath.toLowerCase();
    return tempIndicators.some(indicator => 
        filePathLower.includes(indicator.toLowerCase())
    );
}

/**
 * Get the best available original filename from a source.
 * @param {Object} source - Source object containing metadata
 * @returns {string|null} Original filename if available, null otherwise
 */
export function getOriginalFilename(source) {
    // Priority order for original filename
    const filenameOptions = [
        source.original_filename,
        source.original_name,
        source.display_name,
        source.filename
    ];
    
    // First try to find a non-temp file
    for (const option of filenameOptions) {
        if (option && !isTempFilePath(option)) {
            return option;
        }
    }
    
    // If all options are temp files or null, return the first non-null option
    for (const option of filenameOptions) {
        if (option) {
            return option;
        }
    }
    
    return null;
}

/**
 * Format a source for display, prioritizing original file information.
 * @param {Object} source - Source object containing metadata
 * @returns {string} Formatted string for display
 */
export function formatSourceDisplay(source) {
    // Get the best available original file information
    let originalFile = source.original_filename || 
                      source.original_name || 
                      source.display_name || 
                      source.filename || 
                      source.file_path || '';
    
    if (!originalFile) {
        // Fallback to extracting filename from path
        const filePath = source.file_path || '';
        if (filePath) {
            originalFile = filePath.split(/[/\\]/).pop() || 'Unknown Source';
        } else {
            originalFile = 'Unknown Source';
        }
    }
    
    // Check if this is a temp file
    const isTemp = isTempFilePath(originalFile);
    
    // If it's a temp file, try to get the original name
    if (isTemp) {
        const originalName = source.original_name || source.display_name;
        if (originalName) {
            originalFile = originalName;
        }
    }
    
    // Format the display
    if (originalFile.includes('/') || originalFile.includes('\\')) {
        // For full paths, show a shortened version
        const pathParts = originalFile.replace(/\\/g, '/').split('/');
        if (pathParts.length > 3) {
            const displayPath = `.../${pathParts.slice(-3).join('/')}`;
            return `📄 **${displayPath}**`;
        } else {
            return `📄 **${originalFile}**`;
        }
    } else {
        return `📄 **${originalFile}**`;
    }
}

/**
 * Get source metadata summary.
 * @param {Object} source - Source object containing metadata
 * @returns {Object} Dictionary with key metadata fields
 */
export function getSourceMetadataSummary(source) {
    return {
        original_filename: source.original_filename,
        original_name: source.original_name,
        display_name: source.display_name,
        file_path: source.file_path,
        filename: source.filename,
        upload_source: source.upload_source,
        upload_timestamp: source.upload_timestamp,
        is_temp_file: source.is_temp_file || false,
        source_type: source.source_type,
        doc_id: source.doc_id,
        chunk_id: source.chunk_id
    };
}

/**
 * Format a search result for frontend display with enhanced source information.
 * @param {Object} result - Raw search result from query engine
 * @returns {Object} Formatted result for display
 */
export function formatSearchResultForDisplay(result) {
    // Extract source information
    const sourceInfo = getSourceMetadataSummary(result);
    
    // Format the display
    const formattedResult = {
        content: result.text || '',
        score: result.similarity_score || 0,
        source_display: formatSourceDisplay(result),
        source_info: sourceInfo,
        metadata: result.metadata || {},
        doc_id: result.doc_id,
        chunk_id: result.chunk_id
    };
    
    return formattedResult;
}

/**
 * Upload file with original path information.
 * @param {File} file - File to upload
 * @param {string} originalPath - Original file path (optional)
 * @param {string} description - Additional description (optional)
 * @param {string} uploadSource - Source of the upload (default: web_upload)
 * @param {Object} additionalMetadata - Additional metadata object (optional)
 * @returns {Promise<Object>} Upload response
 */
export async function uploadFileWithOriginalPath(file, originalPath = null, description = null, uploadSource = 'web_upload', additionalMetadata = {}) {
    const formData = new FormData();
    formData.append('file', file);
    
    // Include the original file path if available
    if (originalPath) {
        formData.append('original_path', originalPath);
    } else {
        // Use the file name as fallback
        formData.append('original_path', file.name);
    }
    
    // Add optional fields
    if (description) {
        formData.append('description', description);
    }
    
    if (uploadSource) {
        formData.append('upload_source', uploadSource);
    }
    
    if (Object.keys(additionalMetadata).length > 0) {
        formData.append('additional_metadata', JSON.stringify(additionalMetadata));
    }
    
    const response = await fetch('/api/upload/enhanced', {
        method: 'POST',
        body: formData
    });
    
    if (!response.ok) {
        throw new Error(`Upload failed: ${response.status} ${response.statusText}`);
    }
    
    return response.json();
}

/**
 * Format search results for display with enhanced source information.
 * @param {Array} results - Array of search results
 * @returns {Array} Formatted results for display
 */
export function formatSearchResultsForDisplay(results) {
    return results.map(result => formatSearchResultForDisplay(result));
}

/**
 * Create a source label for display in search results.
 * @param {Object} source - Source object
 * @param {number} index - Index of the source
 * @returns {string} Formatted source label
 */
export function createSourceLabel(source, index) {
    const originalFile = getOriginalFilename(source);
    
    if (originalFile) {
        // If it's a full path, show just the filename
        if (originalFile.includes('/') || originalFile.includes('\\')) {
            const fileName = originalFile.split(/[/\\]/).pop();
            return fileName || `Source ${index + 1}`;
        }
        return originalFile;
    }
    
    return `Source ${index + 1}`;
} 

File: rag_new/rag_system/src/utils/source_formatter.py
"""
Utility functions for formatting source displays with original path information.
"""

from pathlib import Path
from typing import Dict, Any, Optional


def format_source_display(source: Dict[str, Any]) -> str:
    """
    Format a source for display, prioritizing original file information.
    
    Args:
        source: Source dictionary containing metadata
        
    Returns:
        Formatted string for display
    """
    # Get the best available original file information
    original_file = (
        source.get('original_filename') or 
        source.get('original_name') or
        source.get('display_name') or
        source.get('filename') or
        source.get('file_path', '')
    )
    
    if not original_file:
        # Fallback to extracting filename from path
        file_path = source.get('file_path', '')
        if file_path:
            original_file = Path(file_path).name
        else:
            original_file = "Unknown Source"
    
    # Check if this is a temp file
    is_temp = (
        'Temp' in original_file or 
        'tmp' in original_file.lower() or
        '/tmp/' in original_file or
        '\\temp\\' in original_file.lower()
    )
    
    # If it's a temp file, try to get the original name
    if is_temp:
        original_name = source.get('original_name') or source.get('display_name')
        if original_name:
            original_file = original_name
    
    # Format the display
    if '/' in original_file or '\\' in original_file:
        # For full paths, show a shortened version
        path_parts = original_file.replace('\\', '/').split('/')
        if len(path_parts) > 3:
            display_path = f".../{'/'.join(path_parts[-3:])}"
        else:
            display_path = original_file
        return f"📄 **{display_path}**"
    else:
        return f"📄 **{original_file}**"


def get_source_metadata_summary(source: Dict[str, Any]) -> Dict[str, Any]:
    """
    Extract key metadata for source summary.
    
    Args:
        source: Source dictionary containing metadata
        
    Returns:
        Dictionary with key metadata fields
    """
    return {
        'original_filename': source.get('original_filename'),
        'original_name': source.get('original_name'),
        'display_name': source.get('display_name'),
        'file_path': source.get('file_path'),
        'filename': source.get('filename'),
        'upload_source': source.get('upload_source'),
        'upload_timestamp': source.get('upload_timestamp'),
        'is_temp_file': source.get('is_temp_file', False),
        'source_type': source.get('source_type'),
        'doc_id': source.get('doc_id'),
        'chunk_id': source.get('chunk_id')
    }


def format_search_result_for_display(result: Dict[str, Any]) -> Dict[str, Any]:
    """
    Format a search result for frontend display with enhanced source information.
    
    Args:
        result: Raw search result from query engine
        
    Returns:
        Formatted result for display
    """
    # Extract source information
    source_info = get_source_metadata_summary(result)
    
    # Format the display
    formatted_result = {
        'content': result.get('text', ''),
        'score': result.get('similarity_score', 0),
        'source_display': format_source_display(result),
        'source_info': source_info,
        'metadata': result.get('metadata', {}),
        'doc_id': result.get('doc_id'),
        'chunk_id': result.get('chunk_id')
    }
    
    return formatted_result


def is_temp_file_path(file_path: str) -> bool:
    """
    Check if a file path is a temporary file path.
    
    Args:
        file_path: File path to check
        
    Returns:
        True if it's a temp file path, False otherwise
    """
    if not file_path:
        return False
    
    temp_indicators = [
        'Temp',
        'tmp',
        '/tmp/',
        '\\temp\\',
        'tempfile',
        'temp_file'
    ]
    
    file_path_lower = file_path.lower()
    return any(indicator.lower() in file_path_lower for indicator in temp_indicators)


def get_original_filename(source: Dict[str, Any]) -> Optional[str]:
    """
    Get the best available original filename from a source.
    
    Args:
        source: Source dictionary containing metadata
        
    Returns:
        Original filename if available, None otherwise
    """
    # Priority order for original filename
    filename_options = [
        source.get('original_filename'),
        source.get('original_name'),
        source.get('display_name'),
        source.get('filename')
    ]
    
    for option in filename_options:
        if option and not is_temp_file_path(option):
            return option
    
    # If all options are temp files or None, return the first non-None option
    for option in filename_options:
        if option:
            return option
    
    return None 